File: DAA/1-fibonacci.cpp
//  recursive

// #include <iostream>

// int recursiveFibonacci(int n) {
//     if (n <= 1)
//         return n;
//     return recursiveFibonacci(n - 1) + recursiveFibonacci(n - 2);
// }

// int main() {
//     int n = 1; // Change n to the desired Fibonacci number
//     std::cout << "Fibonacci(" << n << ") = " << recursiveFibonacci(n) << std::endl;
//     return 0;
// }


// non-recursive

#include <iostream>

int nonRecursiveFibonacci(int n) {
    if (n <= 1)
        return n;
    
    int a = 0, b = 1, result;
    for (int i = 2; i <= n; i++) {
        result = a + b;
        a = b;
        b = result;
    }
    return result;
}

int main() {
    int n = 10; // Change n to the desired Fibonacci number
    std::cout << "Fibonacci(" << n << ") = " << nonRecursiveFibonacci(n) << std::endl;
    return 0;
}

File: DAA/1-fibonacci.py
#  non recursive fibonacci
# def non_recursive_fibonacci(n):
#     if n <= 1:
#         return n

#     a, b = 0, 1
#     result = 0
#     for i in range(2, n):
#         result = a + b
#         a, b = b, result

#     return result

# if __name__ == "__main__":
#     n = 5
#     print(f"Fibonacci({n}) = {non_recursive_fibonacci(n)}")


# recursive fibonacci

def recursive_fibonacci(n):
    if n <= 1:
        return n
    
    return recursive_fibonacci(n - 1) + recursive_fibonacci(n - 2)

if __name__ == "__main__":
    n = 5
    print(f"Fibonacci({n}) = {recursive_fibonacci(n-1)}")
    